### ISSUE TITLE
Type references count are not shown for const in C#

### ISSUE THREAD

**Issue by eriawan** [dotnet/roslyn#32320]
**Version Used**: VS 2017 15.8.3, 15.9.1, 15.9.5

**Steps to Reproduce**:

1. Create any C# class/code with const
2. Use the const in other class or in other method in the same class
3. Look at the const declaration

**Expected Behavior**:
The const should show how many (count of total) references in the C# project

**Actual Behavior**:
Total references are not shown.

Sample code:
For const:
```csharp
    public class MiscelanousConstants
    {
        public const decimal SPEED_OF_LIGHT = 299792458M;
        public const decimal MATH_PI = 3.14123123M;
        public const decimal ONE_PARSEC = 3.26M;
    }

``` 
Sample use of the const:

```csharp
    class AstronomicDistanceCalculator
    {
        public static decimal CalculateLightYears(decimal yearsFraction)
        {
            return Decimal.Multiply(yearsFraction, Decimal.Multiply(MiscelanousConstants.SPEED_OF_LIGHT, (365 * 24 *3600)));
        }
        public static decimal ConvertParsecToLightYears(decimal parsec)
        {
            return (decimal)(parsec * CalculateLightYears(MiscelanousConstants.ONE_PARSEC));
        }
        public static decimal CalculateLightSeconds(Int32 seconds)
        {
            return Decimal.Multiply((decimal)seconds, MiscelanousConstants.SPEED_OF_LIGHT);
        }
    }
```

```csharp
    class Program
    {
        const decimal ABSOLUTE_ZERO = 0.0M;

        static void Main(string[] args)
        {
            decimal oneLightSecond = MiscelanousConstants.SPEED_OF_LIGHT;
            Console.WriteLine("one light second = " + oneLightSecond + "m");
            Console.WriteLine("one parsec = " + AstronomicDistanceCalculator.ConvertParsecToLightYears(1) + "m");
            Console.WriteLine("zero = " + ABSOLUTE_ZERO);
            Console.ReadLine();
        }
    }
```

On VS 2017, the const declaration has no total of type references:

![image](https://user-images.githubusercontent.com/8773147/50963597-bfe23600-14ff-11e9-92a3-f324df6f2481.png)

Only the enclosing `MiscelanousConstants` class that has the types has total references shown.

But when using "Find All References" on `MiscelanousConstants.SPEED_OF_LIGHT`, I get correct number of the references:

![image](https://user-images.githubusercontent.com/8773147/50963669-ec964d80-14ff-11e9-99f8-8619d5cbdff6.png)

The same problem happens on the const of ABSOLUTE_ZERO:

![image](https://user-images.githubusercontent.com/8773147/50963920-9aa1f780-1500-11e9-8694-c40533a99d31.png)

Finally, my questions are:

1. Is this by design?
2. If is this by design, why const is not showing total references before the declaration, like other type/member such as class, method, field and properties?

If this is not by design and easy to fix, I'd like to propose a PR later.

**Comment by jhinder** [comment]
> like other type/member such as (...) field

Fields don't have Code Lens, and constants are (special) fields, so I'd say it's intentional that constants don't have CodeLens either.

**Comment by eriawan** [comment]
> > like other type/member such as (...) field
> 
> Fields don't have Code Lens, and constants are (special) fields, so I'd say it's intentional that constants don't have CodeLens either.

Is this intention documented in Roslyn or in MS [docs.microsoft.com](https://docs.microsoft.com/dotnet)?
Because if it isn't, imho it's not ok to consider that it's intentional/by design.

**Comment by CyrusNajmabadi** [comment]
This was intentional IIRC.  The major reason was that people just found this too cluttering to have all their fields have all this gunk on them.  For things like methods it was fine given that methods are usually large enough that you're not really impairing much by having a codelens widget on each of them.  On hte other hand, having a whole block of fields now be broken apart was very offputting for many.

That said, this might be somethign that could be done with an opt-in option for CodeLens.

**Comment by jhinder** [comment]
> Is this intention documented

Yes, in several places:

* [C# reference on the `const` keyword](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/const): *You use the `const` keyword to declare a constant field (...)*
* Roslyn: [`IFieldSymbol`](https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/Symbols/IFieldSymbol.cs) has a property `IsConst` that is true *if this field was declared as "const"*.
* [CLI spec](https://www.ecma-international.org/publications/files/ECMA-ST/ECMA-335.pdf): Fields are emitted as `.field [modifier] static literal [type] = [value]`. The spec for the literal constraint says (section I.8.6.1.2): *This constraint (...) shall only be placed on static fields of compound types.*

**Comment by sharwell** [comment]
🔗 Related to #23908

**Comment by eriawan** [comment]
@CyrusNajmabadi 

on this:

> That said, this might be somethign that could be done with an opt-in option for CodeLens.

I agree, because having option to set opt-in for const would be useful, because by default we should not clutter the IDE experiences.

@sharwell 
Is this issue being updated? I mean, is there a consideration to be reviewed first or will it be accepted?

**Comment by Taschenbuch** [comment]
it would be awesome to have this as opted in. I sometimes switched to using properties just to see the references, though i would have liked to use consts instead.

**Comment by freerider7777** [comment]
Having code lens for constants would be helpful for quick code navigation, it's obvious. Maybe there are some technical problems, but "Find All References" finds them, why not code lens.

**Comment by freerider7777** [comment]
Same for enums BTW

**Comment by Guyiming** [comment]
> Having code lens for constants would be helpful for quick code navigation, it's obvious. Maybe there are some technical problems, but "Find All References" finds them, why not code lens.

I Agree

**Comment by ShadowMarker789** [comment]
910 days after the original suggestion was posted, I found myself wanting this feature when trying to work out which public fields and constants of 4000+ line abomination class I'm supposed to be maintaining are not being used. 

Even if it's implemented much later, this would still be a very appreciated addition.

**Comment by CwjXFH** [comment]
@ShadowMarker789 right-click on the constants and click **find all references**, which can help you find where the constants were referenced in your code.

**Comment by dotkiddo** [comment]
Please please, I am desperate for this. 
Heck, if anyone would be willing to give me some guidelines I'd love to implement and PR.
I want this feature for easy reference to consts used in attribute parameters. 
(Because using string properties [even static get/readonly] in attribute parameters are not supported, and I would like a single source of reference truth for a very long list of API routes.)

**Comment by ITheP** [comment]
Be incredibly useful to have for us! Gets our vote.

**Comment by eriawan** [comment]
@dotnet/roslyn 

folks, this issue has been 2 years. I like to have an opt-in to display references on const, and I think this issue has got some up votes.

Is there any update on considerations/reviews on this issue? 
If this marked as "up for grabs" then I would like to submit PR on this one, and I would like to know how to make it optional as opt in.

**Comment by sharwell** [comment]
Fixing this issue (and other cases where CodeLens is desired for code elements that it doesn't consider today) relies on porting and modifying a large amount of internal code to open source. It's not clear when this work would be scheduled, since it would require taking resources away from other active items.

**Comment by eriawan** [comment]
@sharwell 

> Fixing this issue (and other cases where CodeLens is desired for code elements that it doesn't consider today) relies on porting and modifying a relatively large amount of internal code to open source.

OMG... Apologize I didn't know before about the fact that this feature (related to CodeLens) isn't open sourced yet.
I thought all related to Roslyn especially intelisense and CodeLens is opensource, and I also thought if it's available open source part of this repo, I'm willing to help. I'm sorry.

Please let me know if there will be some work allocated in the future for this CodeLens for const support. I think it's quite useful, not just nice to have. My offer to help if it's opensourced and put as "help wanted/up for grabs", still available 🙂

**Comment by sharwell** [comment]
@eriawan it's no problem. I just prefer to answer in as straightforward was as possible if someone's offering help and we aren't in a position it accept it. 😄

**Comment by RFBomb** [comment]
Having this as a toggle-able option would be a great addition. I'm currently re-organizing some code, and had to change all the constants to Get-Only Properties in order to see where they are used and how many times easily.

**Comment by CyrusNajmabadi** [comment]
Design meeting review 4/25/2022.

1. We believe codelens should be enabled on Fields/Constants.
2. There is unlikely to be any significant perf overhead *over and above* what you get if you're already using codelens already.
3. Because of the potential to impact the UI for an experience that has been relatively unchanged for 10+ years, this likely should be paired with an option to allow users to turn this off if they do not like having these new codelens items show up.

We should drive a telemetry effort to see how many people have codelens on, but then disable this new option.  If it turns out that there is practically no userbase that turns this off, we can consider removing the option later.

**Comment by eriawan** [comment]
@CyrusNajmabadi 

This is a great news!
Is there any definite plan on when this feature will be enabled?
I hope we could see this extended CodeLens on constants and fields in the next preview update of VS 2022, the 17.3.0 Preview!

**Comment by CyrusNajmabadi** [comment]
@eriawan no plans as of yet. This is in the backlog currently. The design review just lays out the changes we would take if this gets scheduled for a future sprint. Thanks!

**Comment by WalissonPires** [comment]
+1

**Comment by ashokrathodcg** [comment]
+1

**Comment by Lasko3** [comment]
+1

**Comment by vgdagpin** [comment]
+1

**Comment by CyrusNajmabadi** [comment]
Please don't spam posts. If you want to upvote then you can do so in the op. Thanks!

**Comment by HypsyNZ** [comment]
Is this feature available yet?

**Comment by imniko** [comment]
wow... a simple feature not done even in 4 years...

**Comment by HypsyNZ** [comment]
This thread is proof you don't listen to the community.

**Comment by HypsyNZ** [comment]
>  This is in the backlog currently. 

Yes anything that doesn't directly benefit the circle jerk of cult knowledge gets put on the back burner.

**Comment by HypsyNZ** [comment]
>  This is in the backlog currently. 

Yes anything that doesn't directly benefit the circle jerk of cult knowledge gets put on the back burner so these people can continue to feel smart about things that haven't applied for 15 years.

**Comment by CyrusNajmabadi** [comment]
Locking as this has gotten too heated. As mentioned, we have over 8000 issues from many customers, and a lot of critical things we are working on. This is an open source project and we would welcome help here from interested parties since this currently hasn't been able to make it off the backlog yet.  Thanks!
